macrosynergy/panel/correlation.py

```diff
@@ -0,0 +1,494 @@
+"""
+Functions used to calculate correlation across categories or cross-sections of
+panels.
+"""
+import itertools
+import pandas as pd
+import numpy as np
+from typing import List, Union, Tuple, Dict, Optional
+from collections import defaultdict
+
+from macrosynergy.management.utils import reduce_df
+from macrosynergy.management.simulate import make_qdf
+from macrosynergy.management.utils import _map_to_business_day_frequency
+
+
+def corr(
+    df: pd.DataFrame,
+    xcats: Union[str, List[str]] = None,
+    cids: List[str] = None,
+    xcats_secondary: Optional[Union[str, List[str]]] = None,
+    cids_secondary: Optional[List[str]] = None,
+    start: str = "2000-01-01",
+    end: str = None,
+    val: str = "value",
+    freq: str = None,
+    lags: dict = None,
+    lags_secondary: Optional[dict] = None,
+    return_dates: bool = False,
+) -> Union[pd.DataFrame, Tuple[pd.DataFrame, str, str]]:
+    """
+    Calculate correlation across categories or cross-sections of panels.
+
+    :param <pd.Dataframe> df: standardized JPMaQS DataFrame with the necessary columns:
+        'cid', 'xcats', 'real_date' and at least one column with values of interest.
```

Comment:
`cid`, `xcat` ~`xcats`~, 'real_date' and at least one column with values of interest.